cmd/release-controller/upgrades.go

```diff
@@ -317,34 +322,173 @@ func syncGraphToSecret(graph *UpgradeGraph, update bool, secretClient kv1core.Se
 		}
 		return true, nil
 	}, stopCh)
+}
 
-	if !update {
+func saveUpgradeGraph(buf *bytes.Buffer, graph *UpgradeGraph, secretClient kv1core.SecretInterface, ns, name string) {
+	if err := graph.Save(buf); err != nil {
+		klog.Errorf("Unable to calculate graph state: %v", err)
+		return
+	}
+	secret, err := secretClient.Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Can't read latest secret %s/%s: %v", ns, name, err)
 		return
 	}
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte)
+	}
+	secret.Data["latest"] = buf.Bytes()
+	if _, err := secretClient.Update(context.TODO(), secret, metav1.UpdateOptions{}); err != nil {
+		klog.Errorf("Can't save state to secret %s/%s: %v", ns, name, err)
+	}
+	klog.V(2).Infof("Saved upgrade graph state to %s/%s", ns, name)
+}
 
-	// wait a bit of time to let any other loops load what they can
-	time.Sleep(15 * time.Second)
+func (g *UpgradeGraph) orderedRecords() []UpgradeRecord {
+	records := g.Records()
+	// put the records into a stable order
+	sort.Slice(records, func(i, j int) bool {
+		a, b := records[i], records[j]
+		if a.To == b.To {
+			return a.From < b.From
+		}
+		return a.To < b.To
+	})
+	for _, record := range records {
+		sort.Slice(record.Results, func(i, j int) bool {
+			return record.Results[i].URL < record.Results[j].URL
+		})
+	}
+	return records
+}
 
-	// keep the secret up to date
-	buf := &bytes.Buffer{}
-	wait.Until(func() {
-		buf.Reset()
-		if err := graph.Save(buf); err != nil {
-			klog.Errorf("Unable to calculate graph state: %v", err)
-			return
+const (
+	PruneGraphPrintSecret = "secret"
+	PruneGraphPrintDebug = "debug"
+)
+
+func (c *Controller) pruneGraph(secretClient kv1core.SecretInterface, ns, name string, printOption string, confirm bool) error {
+	stopCh := wait.NeverStop
+
+	loadUpgradeGraph(c.graph, secretClient, ns, name, stopCh)
+
+	imageStreams, err := c.releaseLister.ImageStreams(ns).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var stableTagList []string
+
+	for _, imageStream := range imageStreams {
+		r, ok, err := c.releaseDefinition(imageStream)
+		if err != nil || !ok {
+			continue
 		}
-		secret, err := secretClient.Get(context.TODO(), name, metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("Can't read latest secret %s/%s: %v", ns, name, err)
-			return
+		if r.Config.Name == "4-stable" {
```

Comment:
Was this hardcoded before?  If not, don't use 4-stable, use "As == stable"

Reply:
Yeah, it was.  I fixed it.